README
# **************************************************************************
# FILE : README
#
# LAST MODIFIED : 20 Dec 2007
# ==========================================================================

This software is the cmgui part of CMISS.
It is released under the Mozilla Public License v1.1

Further information, help pages, a tracker and a wiki is available at 
http://www.cmiss.org/cmgui/

This software was originally developed at The University of Auckland and 
the Copyright held by Auckland UniServices Ltd.
This development has been for over 15 years and this release is under ongoing
development.

The strengths of this development are field storage including many finite element
basis types, 3D visualisation and a mathematical field abstraction layer. 

It includes some modules which are being deprecated and phased out, and some
which under active development.
The win32 port is almost an exact copy as the i686-linux cmgui-wx with most of the 
important user interface components included. Active development
The code is being redesigned so that rather than being an overall application
it is a core library used by multiple applications, this includes the starting
of a defined api directory, the building of perl and python interfaces and the
inclusion of the 3D windows in Linux and win32 in a mozilla component.
  


CHANGELOG
/* OpenCMISS-Zinc Library
*
* This Source Code Form is subject to the terms of the Mozilla Public
* License, v. 2.0. If a copy of the MPL was not distributed with this
* file, You can obtain one at http://mozilla.org/MPL/2.0/. */

v3.0.0

API Changes:
3079 Change namespace from zinc {} to OpenCMISS{ Zinc { } }.
3260 Move all enums into types/*id files.
3260 Removed interacetive tool from external APIs.
3260 Removed all execute commands APIs.
3360 Added isValid APIs for all c++ classes.
3554 Removed rendition and merged its functions into scene, scene now is the
	graphical presentation of a region.
3627 Replaced get/set attribute with direct APIs.
3661 Removed graphics module APIs and moved the graphics related modules getter APIs to
	context and scene object.
3662 Changed API to use full type names in create and other functions, 
	and multi-word type names are compounded into a single word e.g. 
	field_cache -> fieldcache (FieldCache -> Fieldcache in C++).
3668 Changed default projection mode to be perspective. Fly zoom mode is now the default right button drag behaviour. 
3678 Added functions in image resample field to specify input coordinates bounds and texture lookup coordinates, using 
	this field will provide a work around to get image filter working on images with texture coordinates outside of the 0 to 1 range.

Added Features:
455 Added stream APIs to support compressed files.
2107 Added glyph APIs, custom glyphs can now be created from glyph module and set on point graphics.
3260 Added setTextureField and setTextureCoordinateField APIs to materials.
3360 Added field, scene filter and glyph cast-to-derived-type methods to C++ API.
3361 Completed graphics APIs, added graphics point attributes APIs for editing point graphics,
	added graphic line attributes APIs for editing line attirbutes.
3362 Added timekeeper module to support adding multiple timekeepers in future.
3425 Added field module notifier and field module event query for individual field changes.
3425 Added mesh changes object and API for enquiring about changes to elements from a field module event.
3454 Sceneviewerinput which handle inputs into libZinc scene viewer which is
	useful for transformation and zooming.
	3498 Adding analyze image format handling.
3528 Reintroduced timekeepers and timenotifier (previously known as time object).
3530 Adding API for Cmiss_field_module_create_field_{derivative,curl,divergence,gradient}.
3532 Adding new API to be able to compute nodeset minimum and nodeset maximum.
3550 Added scene picker which can be used for picking objects from scene with provided information.
3550 Added graphics type scenefilter. 
3581 Added tessellation module for managing tessellation objects.
3583 Added font module for managing font objects.
3586 Added spectrum module for managing spectrum objects.
3589 Added material module for managing material objects.
3597 Added scenefilter module for managing scenefilter objects.
3609 Added scene API to render graphics to a point cloud.
3611 Added more image processing APIs to c++ interface.
3630 Added scene API to get spectrum data range.
3639 Added APIs to set get and set wrap mode of image field.
3646 Added Cmiss_node_set_identifier_and Cmiss_element_set_identifier APIs.
3647 Added new yellow, magenta, cyan, orange, brown, grey25, grey50 and grey75 materials.
3687 Added three new context APIs - get version, get_revision and get_version_string.

Bugs fixed:
Fixed several bugs causing occasional crashes.
3347 Fixed incorrect error message when creating a material

Improvements:
3185 Improved region import/export APIs to support datapoints import and domain specific exports.
3359 Reworked fonts.
3361 Improved efficiency of changing spectrum
3361 Points graphic type with separate field domain type (supersedes use_element_type) to 
	replace point, node_points, data_points, element_points.
3492 Improved FieldImage APIs.
3580 dedicated APIs for get/set is_managed replacing general attribute APIs.
3636 Greatly improved spectrum APIs.
3640 Improved Sceneviewer APIs.

Miscellaneous:
3361 Renamed iso_surfaces to contours.
3406 Change component field to use first component index = 1 (was 0)
3558 Updated license block to MPL2.0.

v2.9.0

API Changes:
(New) Add CMISS_SCENE_VIEWER_INPUT_MODIFIER_NONE mode
(New) Enable more field types for create/define fields APIs

Added Features (some incomplete):
3133 FieldML 0.5 import, replaces FieldML 0.4 import.
3184 Allowed reading data into a specific region.

Bugs fixed:
Fixed several bugs causing occasional crashes.
3115 Key presses now get sent to scene viewer callbacks
3160 Fixed crashes caused when passing a null field to cmzn_field_module_create_abs
3167 Fixed setting linear filter for image field has no effect
3181 Fixed crashes when passing null context to cmzn_context_get_default_region
3207 Fixed streamlines not showing on 2D models

Improvements:
2512 Remove group regions, replaced by group fields
3213 Improved field evaluation/assignment caching

Miscellaneous:
Removed cmgui-motif from current build

v2.8.0

API Changes:
(New) Optimisation API
(New) Graphics material API
(New) Field evaluation and assignment API
(New) Scene graphics filtering API
(New) Cmgui Input/Output (Stream) APIs for regions and image field
(New) Graphics tessellation API
(New)API to support external wx main loop
(New) Image field API
(Break) Remove cmiss_timer API
(Break) Remove texture API

Added Features (some incomplete):
1047 Merge scene and region hierarchies.
1699 Print out antialiasing images using cmgui-wx
1776 Region-capable selection and groups.
1844 Implement fitting and optimisation.
2549 FieldML 0.2 import.
2624 Scene graphics filtering           
2680 Added graphics coordinate system
2683 Export points with wavefront format
2684 Improve spectrum rainbow
2749 Graphics tessellation objects and editor
2775 Add xi discretization modes to streamlines
2842 Read fieldML 0.4 format into Cmgui

Bugs fixed:
Fixed numerous occasional crashes.

source/cell.c
/*******************************************************************************
FILE : cell.c

LAST MODIFIED : 16 January 2001

DESCRIPTION :
Main program for cell.  Based on unemap.
==============================================================================*/
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is cmgui.
 *
 * The Initial Developer of the Original Code is
 * Auckland Uniservices Ltd, Auckland, New Zealand.
 * Portions created by the Initial Developer are Copyright (C) 2005
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */
#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include "cell/cell_interface.h"
#include "cell/cell_window.h"
#include "general/debug.h"
#include "general/error_handler.h"
#include "user_interface/event_dispatcher.h"
#include "user_interface/message.h"
#include "user_interface/user_interface.h"

#include "graphics/glyph.h"
#include "graphics/light.h"
#include "graphics/light_model.h"
#include "graphics/material.h"
#include "graphics/scene.h"
#include "graphics/spectrum.h"
#include "graphics/transform_tool.h"
#include "interaction/interactive_tool.h"
#include "interaction/select_tool.h"
#include "selection/any_object_selection.h"
#include "time/time_keeper.h"

/*
Module constants
----------------
*/
#define CHARSET XmSTRING_DEFAULT_CHARSET
#define MAX_NUMBER_MESSAGES 10

/*
Module variables
----------------
*/
static struct Cell_interface *cell_interface = (struct Cell_interface *)NULL;

/*
Module types
------------
*/

/*
Module functions
----------------
*/
static int display_error_message(char *message,void *dummy)
/*******************************************************************************
LAST MODIFIED : 17 October 2000

DESCRIPTION :
Display a Cell error message.
==============================================================================*/
{
	int return_code;

  ENTER(display_error_message);
  USE_PARAMETER(dummy);
  return_code = printf("ERROR: %s\n",message);
  LEAVE;
  return (return_code);
} /* display_error_message */

static int display_information_message(char *message,void *dummy)
/*******************************************************************************
LAST MODIFIED : 03 February 1999

DESCRIPTION :
Display a Cell information message.
==============================================================================*/
{
	int return_code;

  ENTER(display_error_message);
  USE_PARAMETER(dummy);
  return_code = printf("%s",message);
  LEAVE;
  return (return_code);
} /* display_information_message */

static int display_warning_message(char *message,void *dummy)
/*******************************************************************************
LAST MODIFIED : 03 February 1999

DESCRIPTION :
Display a Cell warning message.
==============================================================================*/
{
	int return_code;

  ENTER(display_warning_message);
  USE_PARAMETER(dummy);
  return_code = printf("Warning: %s\n",message);
  LEAVE;
  return (return_code);
} /* display_warning_message */


/*
Main program
------------
*/
int main(int argc,char *argv[])
/*******************************************************************************
LAST MODIFIED : 17 October 2000

DESCRIPTION :
Main program for cell
==============================================================================*/
{
	int return_code=0;
	float default_light_direction[3]={0.0,-0.5,-1.0};
  struct Colour default_colour,ambient_colour;
	struct GT_object *glyph;
  struct Select_tool *select_tool;
  struct Interactive_tool *transform_tool;
  struct Event_dispatcher *event_dispatcher;
  struct User_interface *user_interface;
  struct Any_object_selection *any_object_selection;
  struct Colour background_colour;
  struct Graphical_material *default_graphical_material;
  struct Light *default_light;
  struct Light_model *default_light_model;
  struct Scene *default_scene;
  struct Spectrum *default_spectrum;
  struct Time_keeper *time_keeper;
  struct LIST(GT_object) *graphics_object_list;
  struct LIST(GT_object) *glyph_list;
	struct MANAGER(Interactive_tool) *interactive_tool_manager;
  struct MANAGER(Light) *light_manager;
  struct MANAGER(Light_model) *light_model_manager;
  struct MANAGER(Graphical_material) *graphical_material_manager;
  struct MANAGER(Scene) *scene_manager;
  struct MANAGER(Spectrum) *spectrum_manager;
  struct MANAGER(Texture) *texture_manager;
  
	ENTER(main);
	/* display the version */
	display_message(INFORMATION_MESSAGE, VERSION "\n");
	/* open the user interface */
	if ((event_dispatcher = CREATE(Event_dispatcher)())
		&&(user_interface = CREATE(User_interface)(&argc, argv, event_dispatcher,
		"Cell", "cell")))
	{
    /* set up messages */
    set_display_message_function(ERROR_MESSAGE,display_error_message,
      (void *)NULL);
    set_display_message_function(INFORMATION_MESSAGE,
      display_information_message,(void *)NULL);
    set_display_message_function(WARNING_MESSAGE,display_warning_message,
      (void *)NULL);

    /* Create the CMGUI objects/managers */
    any_object_selection = CREATE(Any_object_selection)();
    background_colour.red=(float)0;
    background_colour.green=(float)0;
    background_colour.blue=(float)0;
    if (graphical_material_manager = CREATE(MANAGER(Graphical_material))())
    {
      if (default_graphical_material = CREATE(Graphical_material)("default"))
      {
        /* ACCESS so can never be destroyed */
			ACCESS(Graphical_material)(default_graphical_material);
			Cmiss_graphics_material_set_attribute_integer(default_graphical_material, CMISS_GRAPHICS_MATERIAL_ATTRIBUTE_IS_MANAGED, 1);
        if (!ADD_OBJECT_TO_MANAGER(Graphical_material)(
          default_graphical_material,
          graphical_material_manager))
        {
          DEACCESS(Graphical_material)(
            &default_graphical_material);
        }
      }
    }
    if (light_manager = CREATE(MANAGER(Light))())
    {
      if (default_light = CREATE(Light)("default"))
      {
        set_Light_type(default_light,INFINITE_LIGHT);
        default_colour.red=1.0;
        default_colour.green=1.0;
        default_colour.blue=1.0;		
        set_Light_colour(default_light,&default_colour); 
        set_Light_direction(default_light,default_light_direction);
        ACCESS(Light)(default_light);
        if (!ADD_OBJECT_TO_MANAGER(Light)(default_light,light_manager))
        {
          DEACCESS(Light)(&default_light);
        }
      }
    }
    if (light_model_manager = CREATE(MANAGER(Light_model))())
    {
      if (default_light_model = CREATE(Light_model)("default"))
      {
        ambient_colour.red=0.2;
        ambient_colour.green=0.2;
        ambient_colour.blue=0.2;
        Light_model_set_ambient(default_light_model,&ambient_colour);
        Light_model_set_side_mode(default_light_model,
          LIGHT_MODEL_TWO_SIDED);
        ACCESS(Light_model)(default_light_model);
        if (!ADD_OBJECT_TO_MANAGER(Light_model)(
          default_light_model,light_model_manager))
        {
          DEACCESS(Light_model)(&default_light_model);
        }			
      }
    }
    if (glyph_list = CREATE(LIST(GT_object))())
    {
      /* add standard glyphs */
      if (glyph = make_glyph_arrow_line("arrow_line",0.25,0.125))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_arrow_solid("arrow_solid",12,2./3.,1./6.))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_axes("axes",0.1,0.025,0.1))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_cone("cone",12))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_cross("cross"))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_cylinder("cylinder6",6))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_cylinder("cylinder",12))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_cylinder("cylinder_hires",48))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_sphere("diamond",4,2))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_line("line"))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_point("point",g_POINT_MARKER,0))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_sheet("sheet"))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_sphere("sphere",12,6))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
      if (glyph=make_glyph_sphere("sphere_hires",48,24))
      {
        ADD_OBJECT_TO_LIST(GT_object)(glyph,glyph_list);
      }
    }
    graphics_object_list = CREATE(LIST(GT_object))();
    if (spectrum_manager = CREATE(MANAGER(Spectrum))())
    {
			default_spectrum = CREATE(Spectrum)("default");
      if (default_spectrum)
      {
        Spectrum_set_simple_type(default_spectrum,
          BLUE_TO_RED_SPECTRUM);
        Spectrum_set_minimum_and_maximum(default_spectrum,0,1);
        if (!ADD_OBJECT_TO_MANAGER(Spectrum)(default_spectrum,
          spectrum_manager))
        {
          DEACCESS(Spectrum)(&default_spectrum);
        }
      }
    }
    texture_manager = CREATE(MANAGER(Texture))();
    if (scene_manager = CREATE(MANAGER(Scene))())
    {
      if (default_scene = CREATE(Scene)("default"))
      {
        Scene_enable_graphics(default_scene,glyph_list,
          graphical_material_manager,
          default_graphical_material,light_manager,
          spectrum_manager,default_spectrum,
          texture_manager);
        Scene_disable_time_behaviour(default_scene);
        ACCESS(Scene)(default_scene);
        if (!ADD_OBJECT_TO_MANAGER(Scene)(default_scene,scene_manager))
        {
          DEACCESS(Scene)(&default_scene);
        }
      }
    }
    interactive_tool_manager=CREATE(MANAGER(Interactive_tool))();
    transform_tool = CREATE(Interactive_tool_transform)(
      user_interface);
    ADD_OBJECT_TO_MANAGER(Interactive_tool)(transform_tool,
      interactive_tool_manager);
		select_tool = CREATE(Select_tool)(interactive_tool_manager,
			any_object_selection,default_graphical_material,user_interface);
    time_keeper=ACCESS(Time_keeper)(CREATE(Time_keeper)("default",
      event_dispatcher, user_interface));
    
    /* create the cell interface */
    if (cell_interface = CREATE(Cell_interface)(any_object_selection,
      &background_colour,default_graphical_material,default_light,
      default_light_model,default_scene,default_spectrum,time_keeper,
      graphics_object_list,glyph_list,interactive_tool_manager,light_manager,
      light_model_manager,graphical_material_manager,scene_manager,
      spectrum_manager,texture_manager,user_interface,exit_cell_window
#if defined (CELL_DISTRIBUTED)
      ,(struct Element_point_ranges_selection *)NULL
#endif /* defined (CELL_DISTRIBUTED) */
      ))
    {
      /* the GUI loop */
      return_code = application_main_loop(user_interface);
      /* clean-up */
      DESTROY(Cell_interface)(&cell_interface);
      /* reset up messages */
      set_display_message_function(ERROR_MESSAGE,
        (Display_message_function *)NULL, NULL);
      set_display_message_function(INFORMATION_MESSAGE,
        (Display_message_function *)NULL, NULL);
      set_display_message_function(WARNING_MESSAGE,
        (Display_message_function *)NULL, NULL);
      if (select_tool)
      {
        DESTROY(Select_tool)(&select_tool);
      }
      if (transform_tool)
      {
        DESTROY(Interactive_tool)(&transform_tool);
      }
      DESTROY(MANAGER(Interactive_tool))(
        &interactive_tool_manager);
      DEACCESS(Scene)(&default_scene);
      DESTROY(MANAGER(Scene))(&scene_manager);
      DESTROY(Any_object_selection)(&any_object_selection);
      DESTROY(LIST(GT_object))(&graphics_object_list);
      DESTROY(LIST(GT_object))(&glyph_list);
      DEACCESS(Spectrum)(&default_spectrum);
      DESTROY(MANAGER(Spectrum))(&spectrum_manager);
      DEACCESS(Graphical_material)(&default_graphical_material);			
      DESTROY(MANAGER(Graphical_material))(&graphical_material_manager);
      DESTROY(MANAGER(Texture))(&texture_manager);
      DEACCESS(Light_model)(&default_light_model);
      DESTROY(MANAGER(Light_model))(&light_model_manager);
      DEACCESS(Light)(&default_light);
      DESTROY(MANAGER(Light))(&light_manager);
    }
    else
    {
      display_message(ERROR_MESSAGE,"main. "
        "Unable to create Cell interface");
    }
    /* close the user interface */
	 DEACCESS(Time_keeper)(&time_keeper);
	 /* close the user interface */
	 DESTROY(User_interface)(&user_interface);
	 DESTROY(Event_dispatcher)(&event_dispatcher);
  }
  else
  {
    display_message(ERROR_MESSAGE,"Could not open user interface");
    return_code=0;
  }
  LEAVE;
  return(return_code);
} /* main */
